### ISSUE TITLE
Potential skinning issue with glb files

### ISSUE THREAD

**Issue by BalintCsala** [godotengine/godot#84479]
### Godot version

4.1.3

### System information

Windows 11 - Godot 4.1.3 - Forward+

### Issue description

I was experimenting with using Blender's bendy bones for animation by copying the transform of the segments to standard bones, but the exported animation shows up weirdly in Godot.  Here's what the animation at a certain frame should look like
![image](https://github.com/godotengine/godot/assets/33815391/e086f0e6-d7cd-4893-8ec2-98959169ecd9)
And here's the result from Godot:
![image](https://github.com/godotengine/godot/assets/33815391/ffd12572-9615-4b04-b444-9bb9d4f743ab)

The same issue doesn't happen when I export to FBX and use FBX2GLTF and with other GLTF viewers (I tried [this](https://gltf-viewer.donmccurdy.com/) and [this](https://playcanvas.com/viewer)), so this is most likely an edge-case in gltf-s godot doesn't handle properly.

### Steps to reproduce

1. Open the attached project
2. Make sure "test" has "Editable children" enabled
3. Select "AnimationPlayer" and set "Current Animation" to "ArmatureAction"

### Minimal reproduction project

[SkinningIssue.zip](https://github.com/godotengine/godot/files/13259731/SkinningIssue.zip)

**Comment by fire** [comment]
Thats so odd @lyuma someone finally found an error case

**Comment by game009f** [comment]
He was caused by LOD
![1](https://github.com/godotengine/godot/assets/80683955/a7f32213-0d89-4d77-881f-49461513ec92)
![2](https://github.com/godotengine/godot/assets/80683955/61162f7f-e2d3-4e61-b339-e55515960512)

**Comment by BalintCsala** [comment]
Disabling LODs does solve the issue, but should that be on by default in this case?

**Comment by Calinou** [comment]
I suggest [testing](https://docs.godotengine.org/en/latest/contributing/workflow/testing_pull_requests.html) https://github.com/godotengine/godot/pull/84384 locally and see if it resolves the issue on your end.

> Disabling LODs does solve the issue, but should that be on by default in this case?

Yes, LOD isn't supposed to break meshes like this :slightly_smiling_face:

**Comment by fire** [comment]
The lod operator doesn't account for per vertex bone groups. 

tl;dr We only match the closest normals in the vertex attributes.

The layout of inputs are:

```c++
size_t new_index_count = SurfaceTool::simplify_with_attrib_func(
	(unsigned int *)new_indices.ptrw(),
	(const uint32_t *)merged_indices_ptr, index_count,
	merged_vertices_f32.ptr(), merged_vertex_count,
	sizeof(float) * 3, // Vertex stride
	merged_normals_f32.ptr(),
	sizeof(float) * 3, // Attribute stride
	normal_weights, 3,
	index_target,
	max_mesh_error,
	simplify_options,
	&mesh_error);
```

Probably want the notion that bone groups that are close are kept together, but each different bone group has very different values numerically. Like bone group 0 for a "head" is numerically different from bone group 8 for the "neck" even though the vertex groups are connected.

**Comment by lawnjelly** [comment]
Doing LOD on skinned meshes is a fun topic, I'm sure there will be research into this.

Because whether weighted polys can be reduced depends on the poses expected!  :grin: 

At the extreme an animated skeleton with little movement can be decimated as usual. Whereas if just one arm moves a lot, only those polys need be prevented from decimation more.

Simplest scheme is just to not LOD any verts in polys that contain weights. That will preserve the joints but will limit how far poly reduction can take place.

In theory if you have a bunch of poses, you can calculate the max error metrics for use in decimation from these (by software skinning each pose maybe).

**Comment by fire** [comment]
I was investigating using **Smooth Skinning Decomposition with Rigid Bones**. This algorithm extracts Linear Blend Skinning (LBS) from example meshes and can be used for:

- Converting animated mesh sequences to LBS.
- Solving skinning weights from shapes and skeleton poses.
- Solving bone transformations for a mesh animation given skinning weights.

https://github.com/electronicarts/dem-bones

However, it doesn't directly relate to solving this problem of modify our simplification algorithm to better preserve vertices for an animated mesh.

**Comment by zeux** [comment]
The result from the image doesn't really make sense to me, but I'm not sure what the mesh that the simplifier is being fed here is - eg I'm not sure I understand what the deformation structure here is, maybe a lot of vertices are in the same position and only are moved by the final bone deformation. I think the mesh importer skins vertices before simplification but that code could be getting incorrect (default?) transforms in this case maybe.

If the initial bone transforms here are identity and the expectation is that the mesh would be in tact after applying different transforms at runtime, then one possible solution is to extract a single bone id for the dominant bone per vertex (largest weight), and at the minimum use it during vertex merging, and also maybe pass it as an extra attribute with some weight to the simplifier. The update to meshoptimizer makes it so that the attribute count is fully dynamic so there won't be a need to patch the simplifier source code anymore (right now the max attribute count is baked into the library source).

There's research on skinning aware simplification but it's fundamentally a difficult problem to solve in a reasonable amount of time. But it also feels like the image here is more wrong than anything I've seen with normal skinning (I've tested simplifier mostly with pre-deformed mesh vertices and it worked completely fine!).

One issue that was mentioned in https://github.com/godotengine/godot/pull/84384 (btw I don't think that should affect this import one way or the other but not sure) is the unbounded positional error; in effect the simplifier is allowed to produce completely arbitrary result. I do think it would be nice to change this longer term, as it can improve the quality here, but on unskinned meshes I would expect the current approach to be okay if suboptimal - if the deformation is too large, the resulting LOD's error will be large, and it should not be selected for rendering. So I don't think this issue should depend on whether the error is bounded (and I would recommend waiting for bounding the error until Godot fully switches to unmodified meshopt_simplify, so probably 0.21 or thereabouts, as right now just passing a smaller error will not work properly due to differences in metrics that include or exclude attributes in the interface).

**Comment by zeux** [comment]
> the image here is more wrong than anything I've seen with normal skinning

Well, the good news is, I can reproduce this in meshoptimizer's test harness :D so this is not unique to Godot import. And yeah without animation (in the initial position) the simplification looks fine - but that's because the initial pose is a straight cylinder, so the simplifier removes triangles along the cylinder's sides. These collapses are also very low error because, well, the shape doesn't change if you do them.

Of course this all changes once the bones are actually rotated. Something like this is inevitable under bone deformation but this of course is a particularly egregious case.

The only way to solve this particular mesh, other than disabling LOD generation during import :), is to feed *some* information to the simplifier that would tell it something about the deformation. It could be bone id, some info about transforms, or weights, basically anything that helps differentiate the triangles...

One additional issue though is that because the distance error is very small, and distance error is used for LOD selection, this alone won't help enough - it's also the case that a combined error will need to be stored for LOD selection to properly reject other LODs.

For *that* to work, import process would need to store the combined error *and* the default normal weighting will need to be dramatically adjusted - right now the simplification gets weight 2.0 for normals but that dramatically increases the combined position+normal weighting that simplifier uses. I suspect that to make reasonable LOD selection decisions using the combined error, not just position error, on "regular" meshes, the normal weight will need to be in the 0.01..0.1 range, although I haven't really experimented with this.

**Comment by zeux** [comment]
Quick example for incorporating the dominant bone index as an attribute (sorry for video quality, it took me 3 tries to get a sub-10 MB video that apparently GH attachments are limited to :-/):

https://github.com/godotengine/godot/assets/1106629/eb54e2d0-5c96-4dbe-8b2c-5d1f6cf35634

Per comments above, this alone would not suffice because the resulting attribute error is presently ignored so the mesh will have better quality LODs but will still probably display poorly, but this can likely be a reasonable way to attack this based on this. I would probably experiment with more physically-inspired inputs than the index, for example dominant bone translation or something like this, as this will make more sense when integrated into the overall error, but maybe even ID is workable with carefully tuned weighting.

**Comment by fire** [comment]
@zeux I did an attempt to do rig aware lods here and it seems to be ok. https://github.com/V-Sekai/godot/tree/lock-primary-base-influence

Needs optimization

**Comment by bmolyneaux** [comment]
> If we disable LOD generation by default, we lose a lot of its benefits as a lot of people won't bother enabling it manually.

Could we instead disable default LOD generation if there is a skinned mesh present?

**Comment by zeux** [comment]
This issue does not affect most skinned meshes in my experience; the example asset here is an outlier. Most commonly you see skinning applied to characters where the base pose typically has sufficiently complex and curved geometry that the deformation issues are not as significant. It feels more reasonable to disable LOD generation for problematic assets and keep it enabled by default.

**Comment by snowflower** [comment]
I am of the very strong opinion that LoD generation should be off by default.
The default user experience should be 'show me what I made in blender'.

The issue linked below is one of many recurring cases where Auto-Lod setting causes a bad first impression for a new user of Godot.
#106393 

This is a very common issue ( i ran into this issue also when first using godot too!), especially for a new user or someone prototyping with lower polygon or in-progress assets,
where the default import process and editor view of their models from blender shows them a lower LoD model, not the model they made in blender. 

This then causes a lot of wasted time diagnosing the issue and re-importing things to turn off lod.
My direct advice to users ive helped troubleshoot low-polygon / retro polygon modeling,
or armature rigs that are procedurally animated that have a simple rest pose, is to turn off lod.

Another case where lod causes more bug reports and issues:
Vertex Colors not working / giving unpredictable results, due to the default Lod generation setting.
Source meshes with vertex colors to adjust layered texture blending or encoding special values in vertex color RGB for using in VS/Fragment shader isn't possible with the current LoD generation that is on by default.
Again the workaround has been to turn off LoD
#57608


Right now the default lod setting puts alot of extra work on a certain style of model assets ( low poly , vertex coloring, and /or colored not just for visual color but for special blending/vfx, and intentionally limited topology as an aesthetic choice)

A reasonable amount of additional documentation could be part of the onboarding process for a user to just 'turn off lod' if you're going to any of this bullet point list of things I mentioned beforehand. But I still foresee more users reporting bugs in the future about lod/improper asset importing if these concerns are dismissed.

**Comment by snowflower** [comment]
Setting a reasonably high vertex count threshold before default lod generation happens would be a great start that caters to a broad spectrum of assets that get imported. If its a project setting that would be good and reduce a lot of extra ui clicking when importing for some users.
It may be my bias but I see a lot of very low poly/mobile/ps1/ps2 poly budget stuff made with godot as well, I just would like a less aggressive lod generation out of the box for the lower end.

The existing godot documentation for how to optimize a game is good, and mentions using level of detail, so I feel there's adequate pointers for users that are intentionally importing more detailed meshes target current gen consoles or higher end PCs to correct bad performance.
https://docs.godotengine.org/en/stable/tutorials/performance/optimizing_3d_performance.html#level-of-detail-lod

**Comment by zeux** [comment]
Vertex color support can be added fairly easily; it’s a matter of passing an extra few floats to the simplifier. Re: skinning - as I noted above, for many character models the existing code works fine; high poly character skinned models should still have level of detail generated.

**Comment by fire** [comment]
1. switch the metric at runtime (from distance only to combined distance+attribute) to respect attribute deviation.

I'm interested in what this means. This is the closest place I can find https://github.com/godotengine/godot/blob/5e27318b6c81c7eeef462737e6259d78841757e6/servers/rendering/renderer_rd/storage_rd/mesh_storage.h#L472 and there's also a opengl variation of this too.

3. default normal threshold

Yes
